### ISSUE TITLE
Fully dynamic product URL leads to 404 for other pages that matches the number of URL params/segments

### ISSUE THREAD

**Issue by kbalagtey-tacit** [SAP/spartacus#6640]
## Environment Details

- Spartacus:  1.4
- Browser: Chrome
- OS: Mac

## Steps to Reproduce
Configure the product route to be fully dynamic like so
```
ConfigModule.withConfig({
    routing: {
        routes: {
            product: {
                paths: [
                    ':category/:name/:productCode',
                ]
            }
        }
    }
})
```

## Observed Results
Any URL that has 3 segments is considered by Spartacus as a product page even when they are not.

For example, the following routes will result in a 404 because Spartacus thinks they are all product pages.

| URL                          | Actual | Expected      |
|------------------------------|--------|---------------|
| /category/bags/shoulder-bags | 404    | Category Page |
| /cmspage/string/random       | 404    | Content Page  |
| /store-finder/country/US     | 404    | Store Locator |

## Expected Results
There should be a way to have more specific route matchers. 

At a minimum, there should be documentation guideline that having a fully dynamic URL is not possible or recommended.

## Offer to Fix
No

**Comment by Platonn** [comment]
Related issue https://github.com/SAP/cloud-commerce-spartacus-storefront/issues/6583

**Comment by Platonn** [comment]
Let's discuss 3 cases, how we can recognize it's a PDP route:
1. If the PDP route can be recognized by some marker included in the dynamic parameter, a custom URL matcher will help. For example:
> /:category/:name/:productCode-x
> `-x` will be hardcoded and that is how you would know it is a product URL

2. Suppose that we don't have any markers like `-x`. Then the question is: how to discover it's PDP? As Angular `UrlMatcher`s are synchronous, we can leverage them only if we can figure it out *synchronously* (i.e. comparing the 3rd parameter against all possible valid product codes upfront). The codes need to be known upfront, i.e:
- taken from `<meta>` tag in runtime or
- preloadeded from backend endpoint in `APP_INITIALZIER` or
- hardcoded.

3. Suppose we need to call the backend endpoint to verify if a current URL is a PDP (i.e call backend to check if 3rd parameter is a valid product code). It's asynchronous, so URL Matcher cannot be used. Here is an idea of solution, however it needs PoC: We would have to process all routes with `**` route and recognize the page type in the custom guard/resolver (which can be asynchronous). 

@kbalagtey-tacit , which of above is your case?

**Comment by kbalagtey-tacit** [comment]
Thanks, @plabadie ,

I'm glad to see that you are already working on this feature planned for 2.0. 

Our use case is 1 & 3.

For 1, we will make use of the workaround you provided while we wait for the proper solution that will come with 2.0. 

For 3, I was thinking of a similar solution to solve our issue relating to [duplicate pages](https://github.com/SAP/cloud-commerce-spartacus-storefront/issues/5863)

**Comment by Platonn** [comment]
Hi @kbalagtey-tacit 

For 1. Configurable URL matchers were released in version `2.0.0-next.0` in 4a9c3487e2bae06d984517012af9694c89ce2269 Please give it a try. I hope analyzing the URL synchronously suffices.

For 3. Determining the page type (if it's type of ProductPage, CategoryPage or ContantPage) based on the backend call would be very tricky in Spartacus, we haven't tried it yet and it's currently not on our radar. Here is why it's tricky:

Currently Spartacus has a hardcoded logic (in CustomSerializer provided for ngrx RouterStateSerializer) that determines the page type by the presence of the `:productCode` param. The route parameters were defined during the previous phase - synchronous URL matching.

So starting point could be overwriting the logic of the CustomSerializer. But there is much more (i.e. replacing all Spartacus routes with a one `**` route with heavy logic that among other things would call backend to recognize page type. 
 
Tickets that may interest you:
- Make part of CustomSerializer extendable https://github.com/SAP/spartacus/issues/7187
- General routing improvments https://github.com/SAP/spartacus/issues/7070

**Comment by kbalagtey-work** [comment]
Understood, thanks @Platonn for the write-up.
